File: crates/runvault/src/lockfile.rs
//! `.runvault.lock` — how a live run is told apart from one that was killed.
//!
//! `Drop` does not run on SIGKILL or a power cut, so the existence of the lock
//! cannot mean "running": a run that died would stay running forever. The lock
//! carries a heartbeat and the identity of the process, and both are checked.

use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Condvar, Mutex};
use std::thread::JoinHandle;
use std::time::Duration;

use chrono::{DateTime, Local};
use serde::{Deserialize, Serialize};

use crate::error::{Error, Result};
use crate::files;

/// The lock file's name inside a run directory.
pub const LOCK_FILE: &str = ".runvault.lock";

/// How often the heartbeat is refreshed.
///
/// Short enough against the five-minute threshold that missing one refresh does
/// not make a live run look dead.
pub const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(30);

/// How long a heartbeat stays believable.
pub const STALE_AFTER: Duration = Duration::from_secs(300);

/// What `.runvault.lock` holds.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct LockRecord {
    /// The machine the process runs on. A PID only means something there.
    pub host: String,
    /// The process id.
    pub pid: u32,
    /// When that process started, so a recycled PID is not mistaken for it.
    pub process_start_time: Option<u64>,
    /// When the run started.
    pub created_at: String,
    /// Last refresh.
    pub heartbeat_at: String,
}

/// Whether a run without `status.json` is still going.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Liveness {
    /// The process is still there, or the heartbeat is recent enough.
    Running,
    /// Neither holds: the process died without writing `status.json`.
    Stale,
}

/// The start time of a live process, in seconds since the epoch.
pub fn process_start_time(pid: u32) -> Option<u64> {
    let mut system = sysinfo::System::new();
    let target = sysinfo::Pid::from_u32(pid);
    system.refresh_processes(sysinfo::ProcessesToUpdate::Some(&[target]), true);
    system.process(target).map(|p| p.start_time())
}

impl LockRecord {
    /// A record for the current process.
    pub fn for_this_process(created_at: DateTime<Local>) -> Self {
        let pid = std::process::id();
        Self {
            host: crate::env::host(),
            pid,
            process_start_time: process_start_time(pid),
            created_at: created_at.to_rfc3339(),
            heartbeat_at: created_at.to_rfc3339(),
        }
    }

    /// Decides whether the run behind this lock is still alive.
    ///
    /// The PID is only consulted on the machine that wrote the lock; elsewhere
    /// the heartbeat is all there is.
    pub fn liveness(&self, now: DateTime<Local>, this_host: &str) -> Liveness {
        if self.host == this_host
            && let Some(started) = process_start_time(self.pid)
            && Some(started) == self.process_start_time
        {
            return Liveness::Running;
        }
        let fresh = DateTime::parse_from_rfc3339(&self.heartbeat_at)
            .map(|beat| now.signed_duration_since(beat).num_seconds())
            .map(|age| age <= STALE_AFTER.as_secs() as i64)
            .unwrap_or(false);
        if fresh {
            Liveness::Running
        } else {
            Liveness::Stale
        }
    }
}

/// Reads a run's lock file, or `None` when there is none.
pub fn read(run_dir: &Path) -> Result<Option<LockRecord>> {
    let path = run_dir.join(LOCK_FILE);
    match std::fs::read_to_string(&path) {
        Ok(text) => Ok(Some(serde_json::from_str(&text)?)),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(Error::io(path, e)),
    }
}

/// Removes a run's lock file. Missing is success.
pub fn remove(run_dir: &Path) -> Result<()> {
    let path = run_dir.join(LOCK_FILE);
    match std::fs::remove_file(&path) {
        Ok(()) => Ok(()),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(Error::io(path, e)),
    }
}

/// Writes the lock file, replacing it atomically so a reader never sees half of it.
pub fn write(run_dir: &Path, record: &LockRecord) -> Result<()> {
    files::write_json_atomically(&run_dir.join(LOCK_FILE), record)
}

/// The thread that keeps `heartbeat_at` fresh while the run is going.
pub struct Heartbeat {
    stop: Arc<(Mutex<bool>, Condvar)>,
    failed: Arc<AtomicBool>,
    handle: Option<JoinHandle<()>>,
}

impl Heartbeat {
    /// Writes the lock and starts refreshing it every [`HEARTBEAT_INTERVAL`].
    pub fn start(run_dir: &Path, record: LockRecord) -> Result<Self> {
        write(run_dir, &record)?;

        let stop = Arc::new((Mutex::new(false), Condvar::new()));
        let failed = Arc::new(AtomicBool::new(false));
        let path: PathBuf = run_dir.to_path_buf();
        let thread_stop = Arc::clone(&stop);
        let thread_failed = Arc::clone(&failed);

        let handle = std::thread::Builder::new()
            .name("runvault-heartbeat".into())
            .spawn(move || {
                let (lock, cvar) = &*thread_stop;
                let mut record = record;
                let mut guard = lock.lock().expect("heartbeat mutex");
                loop {
                    // The flag is read before waiting as well as after: a stop that
                    // arrives before the first wait would otherwise cost a full
                    // interval before the thread noticed.
                    if *guard {
                        return;
                    }
                    let (next, _) = cvar
                        .wait_timeout(guard, HEARTBEAT_INTERVAL)
                        .expect("heartbeat mutex");
                    guard = next;
                    if *guard {
                        return;
                    }
                    record.heartbeat_at = Local::now().to_rfc3339();
                    if write(&path, &record).is_err() {
                        // The run itself is still valid; only the liveness signal is lost.
                        thread_failed.store(true, Ordering::Relaxed);
                    }
                }
            })
            .map_err(Error::PlainIo)?;

        Ok(Self {
            stop,
            failed,
            handle: Some(handle),
        })
    }

    /// Whether a refresh ever failed. A stopped heartbeat only makes a live run
    /// look dead, so it is reported rather than raised.
    pub fn degraded(&self) -> bool {
        self.failed.load(Ordering::Relaxed)
    }

    /// Stops the thread and waits for it.
    pub fn stop(&mut self) {
        let Some(handle) = self.handle.take() else {
            return;
        };
        {
            let (lock, cvar) = &*self.stop;
            if let Ok(mut stop) = lock.lock() {
                *stop = true;
            }
            cvar.notify_all();
        }
        let _ = handle.join();
    }
}

impl Drop for Heartbeat {
    fn drop(&mut self) {
        self.stop();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn this_process_is_running() {
        let record = LockRecord::for_this_process(Local::now());
        assert_eq!(
            record.liveness(Local::now(), &record.host),
            Liveness::Running
        );
    }

    #[test]
    fn a_dead_process_with_an_old_heartbeat_is_stale() {
        let mut record = LockRecord::for_this_process(Local::now());
        record.pid = 0; // never a live user process
        record.process_start_time = Some(1);
        record.heartbeat_at = (Local::now() - chrono::Duration::hours(2)).to_rfc3339();
        assert_eq!(record.liveness(Local::now(), &record.host), Liveness::Stale);
    }

    #[test]
    fn a_recent_heartbeat_from_another_host_still_counts_as_running() {
        let mut record = LockRecord::for_this_process(Local::now());
        record.host = "another-machine".into();
        assert_eq!(
            record.liveness(Local::now(), "this-machine"),
            Liveness::Running
        );
    }

    #[test]
    fn a_recycled_pid_does_not_revive_a_dead_run() {
        let mut record = LockRecord::for_this_process(Local::now());
        // Same PID, but the process it names started at a different time.
        record.process_start_time = record.process_start_time.map(|t| t + 1);
        record.heartbeat_at = (Local::now() - chrono::Duration::hours(2)).to_rfc3339();
        assert_eq!(record.liveness(Local::now(), &record.host), Liveness::Stale);
    }

    #[test]
    fn the_lock_round_trips_through_the_file() {
        let dir = tempfile::tempdir().unwrap();
        assert!(read(dir.path()).unwrap().is_none());
        let record = LockRecord::for_this_process(Local::now());
        write(dir.path(), &record).unwrap();
        assert_eq!(read(dir.path()).unwrap().as_ref(), Some(&record));
        remove(dir.path()).unwrap();
        assert!(read(dir.path()).unwrap().is_none());
        remove(dir.path()).unwrap(); // removing a missing lock is not an error
    }

    #[test]
    fn the_heartbeat_writes_the_lock_and_cleans_up_after_itself() {
        let dir = tempfile::tempdir().unwrap();
        let mut beat =
            Heartbeat::start(dir.path(), LockRecord::for_this_process(Local::now())).unwrap();
        assert!(dir.path().join(LOCK_FILE).is_file());
        assert!(!beat.degraded());
        beat.stop();
        beat.stop(); // idempotent
    }
}

File: python/src/runvault/lockfile.py
"""`.runvault.lock` — how a live run is told apart from one that was killed.

An interpreter that is killed runs no handler, so the existence of the lock
cannot mean "running": a run that died would stay running forever. The lock
carries a heartbeat and the identity of the process, and both are checked.
"""
from __future__ import annotations

import json
import os
import subprocess
import sys
import threading
import time
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any

from . import files
from .env import host

__all__ = [
    "HEARTBEAT_INTERVAL",
    "Heartbeat",
    "LOCK_FILE",
    "LockRecord",
    "RUNNING",
    "STALE",
    "STALE_AFTER",
    "process_start_time",
    "read",
    "remove",
    "write",
]

#: The lock file's name inside a run directory.
LOCK_FILE = ".runvault.lock"

#: How often the heartbeat is refreshed. Short enough against the five-minute
#: threshold that missing one refresh does not make a live run look dead.
HEARTBEAT_INTERVAL = 30.0

#: How long a heartbeat stays believable.
STALE_AFTER = 300.0

#: The process is still there, or the heartbeat is recent enough.
RUNNING = "running"
#: Neither holds: the process died without writing `status.json`.
STALE = "stale"


def process_start_time(pid: int) -> int | None:
    """When a live process started, in seconds since the epoch.

    `None` when this platform cannot say. A run that cannot prove its process is
    alive falls back to the heartbeat, which is weaker but true; inventing a
    start time would make a recycled PID look like the original process.
    """
    try:
        if sys.platform.startswith("linux"):
            with open(f"/proc/{pid}/stat", encoding="utf-8") as handle:
                fields = handle.read().rpartition(")")[2].split()
            ticks = int(fields[19])  # field 22 of proc(5), counting from the state field
            with open("/proc/stat", encoding="utf-8") as handle:
                boot = next(
                    int(line.split()[1]) for line in handle if line.startswith("btime ")
                )
            return boot + ticks // os.sysconf("SC_CLK_TCK")
        if sys.platform == "darwin":
            printed = subprocess.run(
                ["ps", "-o", "lstart=", "-p", str(pid)],
                capture_output=True, text=True, check=True,
            ).stdout.strip()
            if not printed:
                return None
            return int(time.mktime(time.strptime(printed, "%a %b %d %H:%M:%S %Y")))
    except (OSError, ValueError, StopIteration, subprocess.SubprocessError):
        return None
    return None


@dataclass
class LockRecord:
    """What `.runvault.lock` holds."""

    host: str
    pid: int
    process_start_time: int | None
    created_at: str
    heartbeat_at: str

    @classmethod
    def for_this_process(cls, created_at: datetime | None = None) -> LockRecord:
        """A record for the current process."""
        now = (created_at or datetime.now().astimezone()).isoformat()
        pid = os.getpid()
        return cls(host(), pid, process_start_time(pid), now, now)

    def to_json(self) -> dict[str, Any]:
        return {
            "host": self.host,
            "pid": self.pid,
            "process_start_time": self.process_start_time,
            "created_at": self.created_at,
            "heartbeat_at": self.heartbeat_at,
        }

    def liveness(self, now: datetime, this_host: str) -> str:
        """Decide whether the run behind this lock is still alive.

        The PID is only consulted on the machine that wrote the lock; elsewhere
        the heartbeat is all there is.
        """
        if self.host == this_host and self.process_start_time is not None:
            if process_start_time(self.pid) == self.process_start_time:
                return RUNNING
        try:
            age = (now - datetime.fromisoformat(self.heartbeat_at)).total_seconds()
        except ValueError:
            return STALE
        return RUNNING if age <= STALE_AFTER else STALE


def read(run_dir: Path) -> LockRecord | None:
    """Read a run's lock file, or `None` when there is none."""
    path = Path(run_dir) / LOCK_FILE
    if not path.is_file():
        return None
    return LockRecord(**json.loads(path.read_text(encoding="utf-8")))


def remove(run_dir: Path) -> None:
    """Remove a run's lock file. Missing is success."""
    (Path(run_dir) / LOCK_FILE).unlink(missing_ok=True)


def write(run_dir: Path, record: LockRecord) -> None:
    """Write the lock file atomically, so a reader never sees half of it."""
    files.write_json_atomically(Path(run_dir) / LOCK_FILE, record.to_json())


class Heartbeat:
    """The thread that keeps `heartbeat_at` fresh while the run is going."""

    def __init__(self, run_dir: Path, record: LockRecord) -> None:
        self._dir = Path(run_dir)
        self._record = record
        self._stop = threading.Event()
        self._degraded = False
        self._thread = threading.Thread(
            target=self._beat, name="runvault-heartbeat", daemon=True
        )

    @classmethod
    def start(cls, run_dir: Path, record: LockRecord) -> Heartbeat:
        """Write the lock and start refreshing it every `HEARTBEAT_INTERVAL`."""
        beat = cls(run_dir, record)
        write(run_dir, record)
        beat._thread.start()
        return beat

    def _beat(self) -> None:
        while not self._stop.wait(HEARTBEAT_INTERVAL):
            self._record.heartbeat_at = datetime.now().astimezone().isoformat()
            try:
                write(self._dir, self._record)
            except OSError:
                # The run itself is still valid; only the liveness signal is lost.
                self._degraded = True

    @property
    def degraded(self) -> bool:
        """Whether a refresh ever failed. A stopped heartbeat only makes a live
        run look dead, so it is reported rather than raised."""
        return self._degraded

    def stop(self) -> None:
        """Stop the thread and wait for it. Idempotent."""
        self._stop.set()
        if self._thread.is_alive():
            self._thread.join(timeout=HEARTBEAT_INTERVAL)
